src/session/session_api.c

```diff
@@ -454,49 +454,66 @@ __wt_session_close_internal(WT_SESSION_IMPL *session)
 }
 
 /*
- * __session_reconfigure --
- *     WT_SESSION->reconfigure method.
+ * __session_config_prefetch --
+ *     Configure pre-fetch flags on the session.
  */
 static int
-__session_reconfigure(WT_SESSION *wt_session, const char *config)
+__session_config_prefetch(WT_SESSION_IMPL *session, const char **cfg)
 {
     WT_CONFIG_ITEM cval;
-    WT_DECL_RET;
-    WT_SESSION_IMPL *session;
-
-    session = (WT_SESSION_IMPL *)wt_session;
-    SESSION_API_CALL_PREPARE_NOT_ALLOWED(session, ret, reconfigure, config, cfg);
-    WT_UNUSED(cfg);
 
-    WT_ERR(__wt_txn_context_check(session, false));
-
-    WT_ERR(__wt_session_reset_cursors(session, false));
+    if (S2C(session)->prefetch_auto_on)
+        F_SET(session, WT_SESSION_PREFETCH_ENABLED);
+    else
+        F_CLR(session, WT_SESSION_PREFETCH_ENABLED);
 
     /*
-     * Note that this method only checks keys that are passed in by the application: we don't want
-     * to reset other session settings to their default values.
+     * Override any connection-level pre-fetch settings if a specific session-level setting was
+     * provided.
      */
-    WT_ERR(__wt_txn_reconfigure(session, config));
+    if (__wt_config_gets(session, cfg + 1, "prefetch.enabled", &cval) == 0) {
+        if (cval.val) {
+            if (!S2C(session)->prefetch_available) {
+                F_CLR(session, WT_SESSION_PREFETCH_ENABLED);
+                WT_RET_MSG(session, EINVAL,
+                  "pre-fetching cannot be enabled for the session if pre-fetching is configured as "
+                  "unavailable");
+            } else
+                F_SET(session, WT_SESSION_PREFETCH_ENABLED);
+        } else
+            F_CLR(session, WT_SESSION_PREFETCH_ENABLED);
+    }
 
-    ret = __wt_config_getones(session, config, "ignore_cache_size", &cval);
-    if (ret == 0) {
+    return (0);
+}
+
+/*
+ * __session_config_int --
+ *     Configure basic flags and values on the session. Tested via a unit test.
+ */
+static int
+__session_config_int(WT_SESSION_IMPL *session, const char *config)
+{
+    WT_CONFIG_ITEM cval;
+    WT_DECL_RET;
+
+    if ((ret = __wt_config_getones(session, config, "ignore_cache_size", &cval)) == 0) {
         if (cval.val)
             F_SET(session, WT_SESSION_IGNORE_CACHE_SIZE);
         else
             F_CLR(session, WT_SESSION_IGNORE_CACHE_SIZE);
     }
-    WT_ERR_NOTFOUND_OK(ret, false);
+    WT_RET_NOTFOUND_OK(ret);
 
-    ret = __wt_config_getones(session, config, "cache_cursors", &cval);
-    if (ret == 0) {
+    if ((ret = __wt_config_getones(session, config, "cache_cursors", &cval)) == 0) {
         if (cval.val)
             F_SET(session, WT_SESSION_CACHE_CURSORS);
         else {
             F_CLR(session, WT_SESSION_CACHE_CURSORS);
-            WT_ERR(__session_close_cached_cursors(session));
+            WT_RET(__session_close_cached_cursors(session));
         }
     }
-    WT_ERR_NOTFOUND_OK(ret, false);
+    WT_RET_NOTFOUND_OK(ret);
```

Comment:
The original code tests config in the order ignore_cache_size, cache_cursors, debug.checkpoint_fail_before_turtle_update, debug.release_evict_page, and cache_max_wait_ms. The new __session_config_int() tests in the a different order: ignore_cache_size, cache_cursors, debug.checkpoint_fail_before_turtle_update, cache_max_wait_ms, and debug.release_evict_page. If there are any errors other than WT_NOTFOUND the result is different. Otherwise it does not mater. In any case, is there a reason to change the order?

Reply:
2f051ad88

Reply:
Thank you.